src/OAuth/Repository/User/MongoUserRepository.php

```diff
@@ -0,0 +1,131 @@
+use Pagerfanta\Adapter\DoctrineODMMongoDBAdapter;
+use Pagerfanta\Pagerfanta;
+
+class MongoUserRepository implements UserRepositoryInterface
+{
+    /**
+     * @var DocumentManager
+     */
+    private $manager;
+    /**
+     * @var HasherInterface
+     */
+    private $hasher;
+    /**
+     * @var ObjectRepository
+     */
+    private $repository;
+    /**
+     * @var Pagerfanta
+     */
+    private $paginator;
+
+    public function __construct(DocumentManager $manager, HasherInterface $hasher)
+    {
+        $this->manager = $manager;
+        $this->hasher = $hasher;
+        $this->repository = $this->manager->getRepository('Cerberus:User');
+        $queryBuilder = $this->manager->createQueryBuilder('Cerberus:User');
+        $mongoAdapter = new DoctrineODMMongoDBAdapter($queryBuilder);
+        $this->paginator = new Pagerfanta($mongoAdapter);
+    }
+
+    /**
+     * Get a user entity.
+     *
+     * @param string $username
+     * @param string $password
+     * @param string $grantType The grant type used
+     * @param ClientEntityInterface $clientEntity
+     *
+     * @return UserEntityInterface|false
+     */
+    public function getUserEntityByUserCredentials(
+        $username,
+        $password,
+        $grantType,
+        ClientEntityInterface $clientEntity
+    ) {
+        /** @var User $user */
+        $user = $this->repository->findOneBy(['username' => $username]);
+
+        if (! $user) {
+            return false;
+        }
+
+        if (! $this->hasher->verify($password, $user->getPassword())) {
+            return false;
+        }
+
+        return $user;
+    }
+
+    /**
+     * @param string $id
+     * @return User
+     * @throws EntityNotFoundException
+     */
+    public function find(string $id): User
+    {
+        /** @var User $user */
+        $user = $this->repository->find($id);
+
+        if (! $user) {
+            throw EntityNotFoundException::create(User::class, $id);
+        }
+
+        return $user;
+    }
+
+    /**
+     * @param User $user
+     */
+    public function save(User $user, User ...$users): void
+    {
+        array_unshift($users, $user);
+
+        foreach ($users as $user) {
+            $this->manager->persist($user);
+        }
+
+        $this->manager->flush();
+    }
+
+    /**
+     * @param string $id
+     * @throws EntityNotFoundException
+     */
+    public function delete(string $id): void
+    {
+        $user = $this->find($id);
+
+        $this->manager->remove($user);
+        $this->manager->flush();
+    }
+
+    /**
+     * @param int $page
+     * @param int $perPage
+     * @return PaginatedCollection
+     */
+    public function findPaginated(int $page, int $perPage): PaginatedCollection
+    {
+        $this->paginator->setMaxPerPage($perPage);
+        $this->paginator->setCurrentPage($page);
+        $items = $this->paginator->getCurrentPageResults();
+
+        return new PaginatedCollection($items, $this->paginator);
```

Comment:
It just dawned on me that the signature of the collection constructor can be simplified to just accept the paginator.

As it knows the paginator interface, it is able to get the items (`getCurrentPageResults`) itself. In addition you protect the collection from being able to return inconsistent results. i.e. with the current setup you can pass items which do not exist in the paginator.

Reply:
Sharp! I'll refactor this ASAP!